governance/program/src/instruction.rs

```diff
@@ -0,0 +1,203 @@
+    AddCustomSingleSignerInstruction {
+        /// Slot waiting time between vote period ending and this being eligible for execution
+        hold_up_time: Slot,
+        /// Instruction
+        instruction: Instruction,
+        /// Position in instruction array
+        position: u8,
+    },
+
+    /// [Requires Admin token]
+    /// Remove instruction from the Proposal
+    ///
+    ///   0. `[writable]` Proposal State account
+    ///   1. `[writable]` Proposal instruction account
+    ///   2. `[signer]` Admin account
+    ///   3. `[]` Proposal account
+    ///   4. `[]` Proposal Authority account. PDA with seeds: ['governance',proposal_address]
+    RemoveInstruction,
+
+    /// [Requires Admin token]
+    /// Update instruction hold up time in the Proposal
+    ///
+    ///   0. `[writable]` Proposal instruction account
+    ///   1. `[signer]` Admin account
+    ///   2. `[]` Proposal State account
+    ///   3. `[]` Proposal account
+    ///   4. `[]` Proposal Authority account. PDA with seeds: ['governance',proposal_address]
+    UpdateInstructionHoldUpTime {
+        /// Minimum waiting time in slots for an instruction to be executed after proposal is voted on
+        hold_up_time: Slot,
+    },
+
+    /// [Requires Admin token]
+    /// Cancels Proposal and moves it into Canceled
+    ///
+    ///   0. `[writable]` Proposal state account
+    ///   1. `[writable]` Admin account
+    ///   2. `[]` Proposal account
+    ///   3. `[]` Proposal Authority account. PDA with seeds: ['governance',proposal_address]
+    CancelProposal,
+
+    /// [Requires Signatory token]
+    /// Burns signatory token, indicating you approve of moving this Proposal from Draft state to Voting state
+    /// The last Signatory token to be burned moves the state to Voting
+    ///
+    ///   0. `[writable]` Proposal State account
+    ///   1. `[writable]` Signatory account
+    ///   2. `[writable]` Signatory Mint account
+    ///   3. `[]` Proposal account
+    ///   4. `[signer]` Transfer authority
+    ///   5. `[]` Proposal Authority account. PDA with seeds: ['governance',proposal_address]
+    ///   6. `[]` Token program account.
+    ///   7. `[]` Clock sysvar.
+    SignProposal,
+
+    /// [Requires Voting tokens]
+    ///  Deposits source voting tokens (governance or council), indicating you approve and/or disapprove of running this set of instruction
+    ///  If you tip the consensus then the instructions can begin to be run after their hold up time
+    ///
+    ///   0. `[writable]` Governance Vote Record account. Needs to be set with pubkey set to PDA with seeds of the
+    ///                   1) 'governance' const prefix,
+    ///                   2)  Voter account address
+    ///                   3)  Proposal account address     
+    ///   3. `[]` Proposal account
+    ///   1. `[writable]` Proposal State account
+    ///   2. `[]` Source Token Mint account
+    ///   1. `[writable]` User Source Token account to deposit tokens from
```

Comment:
I'm not sure it's necessary to actually move the source tokens into escrow.  The `Vote` instruction could freeze the tokens in the source account directly: https://github.com/solana-labs/solana-program-library/blob/a539ef021f913a1ccc60ffdec41cd2eb2fc7f299/token/program/src/instruction.rs#L228

`WithdrawSourceTokens` then becomes `ReleaseAccount` or something, that can only be executed once the proposal completes or is cancelled.

Is there a use case for why one would want to vote with less than their total tokens (ie, the `u64` within `Vote::Yes` and `Vote::No`)?  I could also do this myself before voting but transferring the tokens I didn't want to use for voting into a separate token account before calling the `Vote` instruction, so it feels like we don't need this extra knob

Reply:
- It might be simpler to freeze tokens but it would put an unnecessary restriction on the mints which could be used for Governance because once you set freezing authority to None you can't change it.
- Also I'm thinking that once you get more governance tokens you could deposit more to increase your vote weight and equally if you need to you could take them back at any time. If we froze the source account then any further transfers would be more complicated. The assumption being that if you increase/decrease your staked tokens your vote gets adjusted accordingly in all active proposals.
-  I can't think at the moment about any scenario when you wouldn't want to use all your voting weight to vote for a  proposal. The only possible scenario would be if you wanted to use the tokens for something else 

Given the above I would stay with depositing the governing tokens into Governance